### PR TITLE
Reduce metastore query count when fetching statistics

### PR DESCRIPTION
**Pull request by sopel39**
Based on: https://github.com/prestosql/presto/pull/1865/

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 1]
```diff
@@ -168,6 +169,11 @@ private ThriftMetastoreUtil() {}
         return result;
     }
 
+    public static org.apache.hadoop.hive.metastore.api.Table toMetastoreApiTable(TableWithPrivileges table)
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java]: I don't think the other variant of the method is used anymore

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 2]
```diff
@@ -760,7 +728,29 @@ public void revokeRoles(Set<String> roles, Set<HivePrincipal> grantees, boolean
...
     }
 
-    private void invalidatePartitionCache(String databaseName, String tableName)
+    private void invalidateDatabase(String databaseName)
+    {
+        databaseCache.invalidate(databaseName);
+        databaseNamesCache.invalidateAll();
+    }
+
+    protected void invalidateTableWithPartitions(String databaseName, String tableName)
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java]: I don't think this name change is helpful. The method `invalidateTable` is invalidating all data related to a table. The *with partitions* seems strange, since we don't have `TableWithPartitions` object. It's really doing *table and partitions*, but at that point you need to call it *table and partitions and statistics and privileges* (and views, too).

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 3]
```diff
@@ -171,10 +171,12 @@ private ThriftMetastoreUtil() {}
...
     public static org.apache.hadoop.hive.metastore.api.Table toMetastoreApiTable(TableWithPrivileges table)
     {
-        return toMetastoreApiTable(table.getTable(), table.getPrincipalPrivileges());
+        org.apache.hadoop.hive.metastore.api.Table result = toMetastoreApiTableWithoutPrivileges(table.getTable());
+        result.setPrivileges(toMetastoreApiPrincipalPrivilegeSet(table.getPrincipalPrivileges()));
+        return result;
     }
 
-    public static org.apache.hadoop.hive.metastore.api.Table toMetastoreApiTable(Table table, PrincipalPrivileges privileges)
+    public static org.apache.hadoop.hive.metastore.api.Table toMetastoreApiTableWithoutPrivileges(Table table)
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreUtil.java]: We should remove this method since it is only used for tests (squash it into the method above).

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePageSinkMetadataProvider.java [hunk 4]
```diff
@@ -54,7 +54,7 @@ public HivePageSinkMetadataProvider(HivePageSinkMetadata pageSinkMetadata, HiveM
         }
         Optional<Partition> modifiedPartition = modifiedPartitions.get(partitionValues);
         if (modifiedPartition == null) {
-            return delegate.getPartition(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName(), partitionValues);
+            return delegate.getPartition(identity, table.get(), partitionValues);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePageSinkMetadataProvider.java]: can table be `empty`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java [hunk 5]
```diff
@@ -266,7 +266,15 @@ private void alterTable(HiveIdentity identity, String databaseName, String table
...
-        return delegate.getPartition(identity, databaseName, tableName, partitionValues).map(ThriftMetastoreUtil::fromMetastoreApiPartition);
+        Function<org.apache.hadoop.hive.metastore.api.Partition, Partition> fromMetastoreApiPartition = ThriftMetastoreUtil::fromMetastoreApiPartition;
+        boolean isAvroTableWithSchemaSet = delegate.getTable(identity, databaseName, tableName)
+                .map(ThriftMetastoreUtil::isAvroTableWithSchemaSet)
+                .orElse(false);
+        if (isAvroTableWithSchemaSet) {
+            List<FieldSchema> schema = delegate.getFields(identity, databaseName, tableName).get();
+            fromMetastoreApiPartition = partition -> fromMetastoreApiPartition(partition, schema);
+        }
+        return delegate.getPartition(identity, databaseName, tableName, partitionValues).map(fromMetastoreApiPartition);
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java]: Would be great to have a test.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 6]
```diff
@@ -1974,6 +1970,12 @@ private static void renameDirectory(HdfsContext context, HdfsEnvironment hdfsEnv
         }
     }
 
+    private Table getTableOrElseThrow(HiveIdentity identity, String databaseName, String tableName)
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: just call it `getTable`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/TableWithPrivileges.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore;
+
+import static java.util.Objects.requireNonNull;
+
+public class TableWithPrivileges
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/TableWithPrivileges.java]: add toString

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java [hunk 8]
```diff
@@ -25,51 +25,42 @@
...
+import static org.apache.hadoop.hive.common.FileUtils.makePartName;
 
 @Immutable
 public class HivePartitionName
 {
     private final HiveTableName hiveTableName;
-    private final Optional<Table> table; // does not participate in hashCode/equals
     private final List<String> partitionValues;
-    private final Optional<String> partitionName; // does not participate in hashCode/equals
+    private final Optional<Partition> partition; // does not participate in hashCode/equals
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: - isn't `Partition` a much bigger object? Since this is part of the cache, object size matters.
- the class name `HivePartitionName` is no longer appropriate.

Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: > isn't Partition a much bigger object? Since this is part of the cache, object size matters.

`Partition` should be cached in `partitionCache`. Previously we also fetched partitions when obtaining stats, but we didn't cache `Partition` objects. However, we should query only partitions that will participate in query, so they would be fetched anyway.

> the class name HivePartitionName is no longer appropriate.

what name do you suggest?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePageSinkMetadataProvider.java (hunk 4 above)
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePageSinkMetadataProvider.java]: nope., there is if before: `!table.isPresent()`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java (hunk 5 above)
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/BridgingHiveMetastore.java]: this code was part of previous PR. Does not exist anymore

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 6 above)
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: the method would then clash with other `getTable` that does return `Optional<Table>`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java (hunk 8 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HivePartitionName.java]: > what name do you suggest?

`HivePartition`?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java (hunk 6 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: `doGetTable` or something more creative.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java [hunk 9]
```diff
@@ -36,13 +36,13 @@
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/HiveMetastore.java]: I would update commit message to something like:

```
Avoid fetching table information when fetching statistics

Directly fetching table at ThriftHiveMetastore could be expensive
as it bypasses the caching layer.
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java [hunk 10]
```diff
@@ -179,15 +178,15 @@ private static AWSCredentialsProvider getAwsCredentialsProvider(GlueHiveMetastor
...
         }
         if (config.isUseInstanceCredentials()) {
             return InstanceProfileCredentialsProvider.getInstance();
         }
         if (config.getIamRole().isPresent()) {
             return new STSAssumeRoleSessionCredentialsProvider
-                .Builder(config.getIamRole().get(), "presto-session")
-                .build();
+                    .Builder(config.getIamRole().get(), "presto-session")
+                    .build();
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueHiveMetastore.java]: This probably should go to different commit, the one where you already fixing indentation.

**Comment by findepi** [comment]
@sopel39 please rebase, as there were other PRs in this area of code in the meantime.

**Comment by electrum** [comment]
This commit seems wrong:

> Move makePartName to ThriftMetastoreUtil

The method is not specific to Thrift -- it's used by the file and Glue metastores.

**Comment by sopel39** [comment]
Replaced by: https://github.com/prestosql/presto/pull/2184
